cpp/velox/compute/VeloxBackend.cc

```diff
@@ -154,6 +158,64 @@ void VeloxBackend::init(const std::unordered_map<std::string, std::string>& conf
 #endif
 
   std::unordered_map<std::string, std::string> configurationValues;
+#ifdef ENABLE_GCS
+  velox::filesystems::registerGCSFileSystem();
+  std::unordered_map<std::string, std::string> gcsConfig({});
+
+  // https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/master/gcs/CONFIGURATION.md#api-client-configuration
+  std::string gsStorageRootUrl;
+  if (auto got = conf.find("spark.hadoop.fs.gs.storage.root.url"); got != conf.end()) {
+    gsStorageRootUrl = got->second;
+  }
+  if (!gsStorageRootUrl.empty()) {
+    std::string gcsScheme;
+    std::string gcsEndpoint;
+
+    const auto sep = std::string("://");
+    const auto pos = gsStorageRootUrl.find_first_of(sep);
+    if (pos != std::string::npos) {
+      gcsScheme = gsStorageRootUrl.substr(0, pos);
+      gcsEndpoint = gsStorageRootUrl.substr(pos + sep.length());
+    }
+
+    if (!gcsEndpoint.empty() && !gcsScheme.empty()) {
+      gcsConfig.insert({"hive.gcs.scheme", gcsScheme});
+      gcsConfig.insert({"hive.gcs.endpoint", gcsEndpoint});
+    }
+  }
+
+  // https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/master/gcs/CONFIGURATION.md#authentication
+  std::string gsAuthType;
+  if (auto got = conf.find("spark.hadoop.fs.gs.auth.type"); got != conf.end()) {
+    gsAuthType = got->second;
+  }
+  if (gsAuthType == "SERVICE_ACCOUNT_JSON_KEYFILE") {
+    std::string gsAuthServiceAccountJsonKeyfile;
+    if (auto got = conf.find("spark.hadoop.fs.gs.auth.service.account.json.keyfile"); got != conf.end()) {
+      gsAuthServiceAccountJsonKeyfile = got->second;
+    }
+
+    std::string gsAuthServiceAccountJson;
+    if (!gsAuthServiceAccountJsonKeyfile.empty()) {
+      auto stream = std::ifstream(gsAuthServiceAccountJsonKeyfile);
+      stream.exceptions(std::ios::badbit);
+      gsAuthServiceAccountJson = std::string(std::istreambuf_iterator<char>(stream.rdbuf()), {});
+    }
+    else {
+      LOG(WARNING) << "STARTUP: conf spark.hadoop.fs.gs.auth.type is set to SERVICE_ACCOUNT_JSON_KEYFILE, "
```

Comment:
If GCS is enabled and auth type is set to service account json keyfile, do we want to throw an exception if json key file is not set rather than just warning? The GCS calls will anyways fail

Reply:
Agree, I will add code to raise exceptions.